tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -220,49 +426,59 @@ static StatusOr<absl::optional<se::blas::AlgorithmType>> DoUncachedGemmAutotune(
 static StatusOr<absl::optional<se::blas::AlgorithmType>> DoGemmAutotune(
     const HloInstruction* instr, const GemmBackendConfig& gemm_config,
     se::DeviceMemoryAllocator* allocator, se::Stream* stream) {
+  VLOG(3) << "Starting autotune of GemmThunk " << instr->ToString();
   const HloInstruction* lhs = instr->operand(0);
   const HloInstruction* rhs = instr->operand(1);
 
+  GpuGemmConfig config = GetGpuGemmConfig(instr);
+  PrimitiveType element_type = config.output_shape.element_type();
   // Don't run autotuning concurrently on the same GPU.
   absl::MutexLock gpu_lock(&GetGpuMutex(stream->parent()));
 
-  GemmCacheKey key =
-      std::make_tuple(stream->parent(), lhs->shape(), rhs->shape(),
-                      instr->shape(), gemm_config.SerializeAsString());
+  if (stream->parent()->SupportsBlasPlans() && config.use_cublaslt &&
+      BlasPlansCompatibleType(element_type)) {
+    TF_RETURN_IF_ERROR(
+        DoBlasPlansAutotune(stream, instr, allocator, gemm_config));
+    return {se::blas::kNoAlgorithm};
```

Comment:
What is the significance of returning `kNoAlgorithm` here?

Reply:
The cuBLASLt autotuner operates on `se::blas::AlgorithmConfig` instead of `se::blas::AlgorithmType` used in non-cuBLASLt autotuning. As such, the result of cuBLASLt autotuning is incompatible with the return value of `DoGemmAutotune` and the `se::blas::kNoAlgorithm` dummy value is returned instead. The outcome of cuBLASLt autotuning is stored in the instance of `BlasPlansAutotuneCacheSingleton`.